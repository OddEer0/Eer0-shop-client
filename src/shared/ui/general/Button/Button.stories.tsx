import { ComponentMeta, ComponentStory } from "@storybook/react"
import styled from "styled-components"

import { Button } from "./Button"

export default {
	title: "UI/Button",
	component: Button
} as ComponentMeta<typeof Button>

const Template: ComponentStory<typeof Button> = args => <Button {...args} />

const Line = styled.div`
	width: 100%;
	height: 100px;
`

export const Default = Template.bind({})
Default.args = {}

export const Variant = () => (
	<>
		<Button>Text</Button>
		<Button variant="contained">Contained</Button>
		<Button variant="outlined">Outlined</Button>
	</>
)

export const Color = () => (
	<>
		<Button>Primary</Button>
		<Button color="secondary">Secondary</Button>
		<Button color="success">Success</Button>
		<Button color="danger">Danger</Button>
		<Button color="warning">Warning</Button>
		<Line />
		<Button variant="outlined">Primary</Button>
		<Button variant="outlined" color="secondary">
			Secondary
		</Button>
		<Button variant="outlined" color="success">
			Success
		</Button>
		<Button variant="outlined" color="danger">
			Danger
		</Button>
		<Button variant="outlined" color="warning">
			Warning
		</Button>
		<Line />
		<Button variant="contained">Primary</Button>
		<Button variant="contained" color="secondary">
			Secondary
		</Button>
		<Button variant="contained" color="success">
			Success
		</Button>
		<Button variant="contained" color="danger">
			Danger
		</Button>
		<Button variant="contained" color="warning">
			Warning
		</Button>
	</>
)

export const Size = () => (
	<>
		<Button variant="contained" size="small">
			Small
		</Button>
		<Button variant="contained">Medium</Button>
		<Button variant="contained" size="large">
			Large
		</Button>
	</>
)
