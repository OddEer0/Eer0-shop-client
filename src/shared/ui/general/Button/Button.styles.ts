/* eslint-disable @typescript-eslint/no-unused-vars */
import styled from "styled-components"

export const Btn = styled.button`
	&.prefix {
		cursor: pointer;
	}

	&.small {
		border-radius: 6px;
		padding: 6px 10px;
		font-size: 16px;
	}

	&.medium {
		border-radius: 10px;
		padding: 8px 12px;
		font-size: 18px;
	}

	&.large {
		border-radius: 12px;
		padding: 10px 14px;
		font-size: 20px;
	}

	&.text {
		&.primary {
			color: ${({ theme }) => theme.ui.text.primary_50};
			background: transparent;
			transition: all 0.3s ease-in-out;

			&:hover {
				background: ${({ theme }) => theme.ui.color.default_50};
				color: ${({ theme }) => theme.ui.text.primary_60};
			}

			&:active {
				background: ${({ theme }) => theme.ui.color.danger_70};
				color: ${({ theme }) => theme.ui.text.primary_70};
			}
		}

		&.secondary {
			color: ${({ theme }) => theme.ui.text.secondary_50};
			background: transparent;
			transition: all 0.3s ease-in-out;

			&:hover {
				background: ${({ theme }) => theme.ui.color.default_50};
				color: ${({ theme }) => theme.ui.text.secondary_60};
			}

			&:active {
				background: ${({ theme }) => theme.ui.color.danger_70};
				color: ${({ theme }) => theme.ui.text.secondary_70};
			}
		}

		&.danger {
			color: ${({ theme }) => theme.ui.text.danger_50};
			background: transparent;
			transition: all 0.3s ease-in-out;

			&:hover {
				background: ${({ theme }) => theme.ui.color.default_50};
				color: ${({ theme }) => theme.ui.text.danger_60};
			}

			&:active {
				background: ${({ theme }) => theme.ui.color.danger_70};
				color: ${({ theme }) => theme.ui.text.danger_70};
			}
		}

		&.success {
			color: ${({ theme }) => theme.ui.text.success_50};
			background: transparent;
			transition: all 0.3s ease-in-out;

			&:hover {
				background: ${({ theme }) => theme.ui.color.default_50};
				color: ${({ theme }) => theme.ui.text.success_60};
			}

			&:active {
				background: ${({ theme }) => theme.ui.color.danger_70};
				color: ${({ theme }) => theme.ui.text.success_70};
			}
		}

		&.warning {
			color: ${({ theme }) => theme.ui.text.warning_50};
			background: transparent;
			transition: all 0.3s ease-in-out;

			&:hover {
				background: ${({ theme }) => theme.ui.color.default_50};
				color: ${({ theme }) => theme.ui.text.warning_60};
			}

			&:active {
				background: ${({ theme }) => theme.ui.color.danger_70};
				color: ${({ theme }) => theme.ui.text.warning_70};
			}
		}
	}

	&.contained {
		&.primary {
			color: ${({ theme }) => theme.ui.text.primary_50};
			background: ${({ theme }) => theme.ui.color.primary_50};
			transition: all 0.3s ease-in-out;

			&:hover {
				background: ${({ theme }) => theme.ui.color.primary_60};
				color: ${({ theme }) => theme.ui.text.primary_60};
			}

			&:active {
				background: ${({ theme }) => theme.ui.color.primary_70};
				color: ${({ theme }) => theme.ui.text.primary_70};
			}
		}

		&.secondary {
			color: ${({ theme }) => theme.ui.text.secondary_50};
			background: ${({ theme }) => theme.ui.color.secondary_50};
			transition: all 0.3s ease-in-out;

			&:hover {
				background: ${({ theme }) => theme.ui.color.secondary_60};
				color: ${({ theme }) => theme.ui.text.secondary_60};
			}

			&:active {
				background: ${({ theme }) => theme.ui.color.secondary_70};
				color: ${({ theme }) => theme.ui.text.secondary_70};
			}
		}

		&.danger {
			color: ${({ theme }) => theme.ui.text.danger_50};
			background: ${({ theme }) => theme.ui.color.danger_50};
			transition: all 0.3s ease-in-out;

			&:hover {
				background: ${({ theme }) => theme.ui.color.danger_60};
				color: ${({ theme }) => theme.ui.text.danger_60};
			}

			&:active {
				background: ${({ theme }) => theme.ui.color.danger_70};
				color: ${({ theme }) => theme.ui.text.danger_70};
			}
		}

		&.success {
			color: ${({ theme }) => theme.ui.text.danger_50};
			background: ${({ theme }) => theme.ui.color.danger_50};
			transition: all 0.3s ease-in-out;

			&:hover {
				background: ${({ theme }) => theme.ui.color.danger_60};
				color: ${({ theme }) => theme.ui.text.danger_60};
			}

			&:active {
				background: ${({ theme }) => theme.ui.color.danger_70};
				color: ${({ theme }) => theme.ui.text.danger_70};
			}
		}

		&.warning {
			color: ${({ theme }) => theme.ui.text.warning_50};
			background: ${({ theme }) => theme.ui.color.warning_50};
			transition: all 0.3s ease-in-out;

			&:hover {
				background: ${({ theme }) => theme.ui.color.warning_60};
				color: ${({ theme }) => theme.ui.text.warning_60};
			}

			&:active {
				background: ${({ theme }) => theme.ui.color.warning_70};
				color: ${({ theme }) => theme.ui.text.warning_70};
			}
		}
	}

	&.outlined {
		border: 2px solid;

		&.primary {
			color: ${({ theme }) => theme.ui.text.primary_50};
			border-color: ${({ theme }) => theme.ui.color.primary_50};
			background: transparent;
			transition: all 0.3s ease-in-out;

			&:hover {
				background: ${({ theme }) => theme.ui.color.default_50};
				border-color: ${({ theme }) => theme.ui.color.primary_60};
				color: ${({ theme }) => theme.ui.text.primary_60};
			}

			&:active {
				background: ${({ theme }) => theme.ui.color.danger_70};
				border-color: ${({ theme }) => theme.ui.color.primary_70};
				color: ${({ theme }) => theme.ui.text.primary_70};
			}
		}

		&.secondary {
			color: ${({ theme }) => theme.ui.text.secondary_50};
			border-color: ${({ theme }) => theme.ui.color.secondary_50};
			background: transparent;
			transition: all 0.3s ease-in-out;

			&:hover {
				background: ${({ theme }) => theme.ui.color.default_50};
				border-color: ${({ theme }) => theme.ui.color.secondary_60};
				color: ${({ theme }) => theme.ui.text.secondary_60};
			}

			&:active {
				background: ${({ theme }) => theme.ui.color.danger_70};
				border-color: ${({ theme }) => theme.ui.color.secondary_70};
				color: ${({ theme }) => theme.ui.text.secondary_70};
			}
		}

		&.danger {
			color: ${({ theme }) => theme.ui.text.danger_50};
			border-color: ${({ theme }) => theme.ui.color.danger_50};
			background: transparent;
			transition: all 0.3s ease-in-out;

			&:hover {
				background: ${({ theme }) => theme.ui.color.default_50};
				border-color: ${({ theme }) => theme.ui.color.danger_60};
				color: ${({ theme }) => theme.ui.text.danger_60};
			}

			&:active {
				background: ${({ theme }) => theme.ui.color.danger_70};
				border-color: ${({ theme }) => theme.ui.color.danger_70};
				color: ${({ theme }) => theme.ui.text.danger_70};
			}
		}

		&.success {
			color: ${({ theme }) => theme.ui.text.success_50};
			border-color: ${({ theme }) => theme.ui.color.success_50};
			background: transparent;
			transition: all 0.3s ease-in-out;

			&:hover {
				background: ${({ theme }) => theme.ui.color.default_50};
				border-color: ${({ theme }) => theme.ui.color.success_60};
				color: ${({ theme }) => theme.ui.text.success_60};
			}

			&:active {
				background: ${({ theme }) => theme.ui.color.danger_70};
				border-color: ${({ theme }) => theme.ui.color.success_70};
				color: ${({ theme }) => theme.ui.text.success_70};
			}
		}

		&.warning {
			color: ${({ theme }) => theme.ui.text.warning_50};
			border-color: ${({ theme }) => theme.ui.color.warning_50};
			background: transparent;
			transition: all 0.3s ease-in-out;

			&:hover {
				background: ${({ theme }) => theme.ui.color.default_50};
				border-color: ${({ theme }) => theme.ui.color.warning_60};
				color: ${({ theme }) => theme.ui.text.warning_60};
			}

			&:active {
				background: ${({ theme }) => theme.ui.color.danger_70};
				border-color: ${({ theme }) => theme.ui.color.warning_70};
				color: ${({ theme }) => theme.ui.text.warning_70};
			}
		}
	}
`
