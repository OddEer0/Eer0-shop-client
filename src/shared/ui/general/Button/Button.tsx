/* eslint-disable import/order */
import { FC, PropsWithChildren } from "react"

import { cn } from "@/shared/helpers"

import { Btn } from "./Button.styles"
import { ButtonProps } from "./Button.types"

export const Button: FC<PropsWithChildren<ButtonProps>> = ({
	isLoading = false,
	size = "medium",
	color = "primary",
	variant = "text",
	className = "",
	children,
	...props
}) => {
	return (
		<Btn className={cn("prefix", variant, size, color, className)} {...props}>
			{isLoading ? <p>loading</p> : children}
		</Btn>
	)
}
