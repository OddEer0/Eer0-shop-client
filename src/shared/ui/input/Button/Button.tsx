/* eslint-disable import/order */
import { FC, PropsWithChildren } from "react"

import { cn } from "@/shared/helpers"

import { ButtonProps } from "./Button.types"
import { ContainedButton } from "./styles/contained.styles"
import { OutlinedButton } from "./styles/outlined.styles"
import { TextButton } from "./styles/text.styles"

export const Button: FC<PropsWithChildren<ButtonProps>> = ({
	size = "medium",
	color = "primary",
	variant = "text",
	className = "",
	children,
	...props
}) => {
	const cl = cn("prefix", variant, size, color, className)

	return (
		<>
			{variant === "contained" ? (
				<ContainedButton className={cl} {...props}>
					{children}
				</ContainedButton>
			) : variant === "outlined" ? (
				<OutlinedButton className={cl} {...props}>
					{children}
				</OutlinedButton>
			) : (
				<TextButton className={cl} {...props}>
					{children}
				</TextButton>
			)}
		</>
	)
}
