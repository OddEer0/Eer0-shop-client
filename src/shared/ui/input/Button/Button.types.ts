import { HTMLAttributes } from "react"

type VariantType = "outlined" | "contained" | "text"
type ColorType = "primary" | "secondary" | "tertiary" | "quaternary" | "danger" | "warning" | "success" | "default"
type SizeType = "small" | "medium" | "large"

export interface ButtonProps extends HTMLAttributes<HTMLButtonElement> {
	variant?: VariantType
	color?: ColorType
	size?: SizeType
	isLoading?: boolean
	className?: string
}
