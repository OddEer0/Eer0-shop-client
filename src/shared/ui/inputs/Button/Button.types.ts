import { HTMLAttributes } from "react"

type VariantType = "outlined" | "contained" | "text"
type ColorType = "primary" | "secondary" | "danger" | "warning" | "success"
type SizeType = "small" | "medium" | "large"

export interface ButtonProps extends HTMLAttributes<HTMLButtonElement> {
	variant?: VariantType
	color?: ColorType
	size?: SizeType
	isLoading?: boolean
	className?: string
}
