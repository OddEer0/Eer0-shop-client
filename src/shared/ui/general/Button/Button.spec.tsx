/* eslint-disable @typescript-eslint/no-explicit-any */
import { shallow } from "enzyme"

import { Button } from "./Button"

describe("Рендер кнопки", () => {
	let component: any
	beforeEach(() => {
		component = shallow(<Button />)
	})

	it("Проверка на рендер кнопки", () => {
		const button = component.find("button")
		expect(button.length).toBe(1)
	})

	it("Рендер contained кнопки", () => {
		const component = shallow(<Button variant="contained" />)
		const wrapper = component.find(".contained")
		expect(wrapper.length).toBe(1)
	})

	it("Snapshot кнопки", () => {
		const component = shallow(<Button variant="contained">BUTTON</Button>)
		expect(component).toMatchSnapshot()
	})
})
