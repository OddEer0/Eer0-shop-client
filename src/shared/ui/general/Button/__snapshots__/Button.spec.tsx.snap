// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Рендер кнопки Snapshot кнопки 1`] = `
<button
  className="prefix contained medium primary "
>
  BUTTON
</button>
`;
